pkg/test/framework/components/echo/docker/images/image.go

```diff
@@ -0,0 +1,279 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package images
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+
+	"istio.io/istio/pkg/test/docker"
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/envoy"
+	"istio.io/istio/pkg/test/framework/components/environment/native"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/crosscompile"
+)
+
+const (
+	noSidecarDockerFile = "pkg/test/echo/docker/Dockerfile.app"
+	sidecarDockerFile   = "pkg/test/echo/docker/Dockerfile.app_sidecar"
+	sidecarStartScript  = "pkg/test/echo/docker/echo-start.sh"
+	certsDir            = "tests/testdata/certs"
+	certFile            = certsDir + "/cert.crt"
+	keyFile             = certsDir + "/cert.key"
+	istioDockerDir      = "docker"
+	caCertsZipFile      = istioDockerDir + "/ca-certificates.tgz"
+	echoClientDir       = "pkg/test/echo/cmd/client"
+	echoServerDir       = "pkg/test/echo/cmd/server"
+	pilotAgentDir       = "pilot/cmd/pilot-agent"
+	nodeAgentDir        = "security/cmd/node_agent"
+	packagingDir        = "tools/packaging"
+	packagingCommonDir  = packagingDir + "/common"
+	postinstScript      = packagingDir + "/deb/postinst.sh"
+	noSidecarImageName  = "app"
+	sidecarImageName    = "app_sidecar"
+)
+
+var (
+	sidecarBuilder   *docker.ImageBuilder
+	noSidecarBuilder *docker.ImageBuilder
+	initMutex        sync.Mutex
+	prebuilt         *Instance
+)
+
+func init() {
+	if env.TAG.Value() != "" {
+		hubPrefix := ""
+		if env.HUB.Value() != "" {
+			hubPrefix = env.HUB.Value() + "/"
+		}
+
+		prebuilt = &Instance{
+			NoSidecar: docker.Image(fmt.Sprintf("%s%s:%s", hubPrefix, noSidecarImageName, env.TAG.Value())),
+			Sidecar:   docker.Image(fmt.Sprintf("%s%s:%s", hubPrefix, sidecarImageName, env.TAG.Value())),
+		}
+
+		scopes.CI.Infof("[Echo Docker]: Using pre-built Docker images: %+v", *prebuilt)
```

Comment:
Do we have these images published anywhere already?

Reply:
Yes ... they're already available in CI.  `Dockerfile.app` is the only one currently used.  I added `Dockerfile.app_sidecar` in #15136.

Reply:
I just wasnt sure if the are published in our release process. `docker pull gcr.io/istio-release/app_sidecar:master-latest-daily` suggest it is though which is nice

Reply:
Yeah, #15136 adds it to the flow for all of the aggregate docker targets (e.g. `make docker push`)